### ISSUE TITLE
poor error message when casting from fat raw pointer to usize

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#27450]
[playpen](https://play.rust-lang.org/?code=trait%20Num%20%7B%7D%0Aimpl%20Num%20for%20i32%20%7B%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20%260i32%20as%20*const%20Num%20as%20usize%3B%0A%7D%0A&version=beta)

```rust
trait Num {}
impl Num for i32 {}

fn main() {
    &0i32 as *const Num as usize;
}
```

This compiles on stable, but beta and nightly require an extra `as *const ()` in the middle to cast away the fat. The error message says to use a "raw pointer" which is misleading as `*const Num` is already a raw pointer, it's just a fat one.

Since this code compiles on stable, I'm not sure if it's also a regression or if I missed an RFC (to be fair, I miss a lot of RFCs).

**Comment by bluss** [comment]
I'm sure it went in as a bug fix. cc @arielb1

**Comment by arielb1** [comment]
This is indeed intentional: the spec at https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md has this relevant part: `e` has type `*T` and `U` is a numeric type, while `T: Sized`; *ptr-addr-cast*. IIRC @nrc made that decision.

**Comment by arielb1** [comment]
The old (1.1) cast code was a badly specified mess of if-s. Do not report bugs.

**Comment by durka** [comment]
Noted. Is this still a (minor) diagnostics issue, or I just don't
understand the terminology? The compiler's advice is to cast through a "raw
pointer", but the actual fix is to cast through a _thin_ pointer.

**Comment by arielb1** [comment]
This may be a diagnostic issue.
